src/libs/relay/conduit_relay_io_silo.cpp

```diff
@@ -411,6 +411,269 @@ namespace silo
+        }
+    }
+
+    std::string
+    Name(int idx) const
+    {
+        std::string res = "";
+
+        // bounds check
+        if (idx < 0 || idx >= nblocks)
+        {
+            return res;
+        }
+
+        // check for simple case
+        if (0 < names_list.size())
+        {
+            return names_list[idx];
+        }
+
+        if (nullptr != empty_list)
+        {
+            int bot = 0;
+            int top = empty_count - 1;
+            int mid;
+            while (bot <= top)
+            {
+                mid = (bot + top) >> 1;
+
+                if (idx > empty_list[mid])
+                {
+                    bot = mid + 1;
+                }
+                else if (idx < empty_list[mid])
+                {
+                    top = mid - 1;
+                }
+                else
+                {
+                    return "EMPTY";
+                }
+            }
+        }
+
+        // namescheme case
+        if (nullptr != file_namescheme)
+        {
+            const char *file_res = DBGetName(file_namescheme, idx);
+            if (nullptr != file_res)
+            {
+                res += (std::string(file_res) + ":");
+            }
+        }
+
+        if (nullptr != block_namescheme)
+        {
+            const char *block_res = DBGetName(block_namescheme, idx);
+            if (nullptr != block_res)
+            {
+                res += std::string(block_res);
+            }
+        }
+
+        return res;
+    }
+};
+
+//-----------------------------------------------------------------------------
+// fetches the paths from part of the root silo index
+std::vector<std::string> 
+get_paths(const Node &silo_index_path,
+          const std::string &path_string,
+          const int num_domains)
+{
+    std::vector<std::string> paths;
+    if (silo_index_path.has_child(path_string))
+    {
+        // Preallocate memory to avoid repeated reallocations
+        paths.reserve(num_domains);
+        for (int block_id = 0; block_id < num_domains; block_id ++)
+        {
+            //                mesh_index   ["mesh_paths"][domain#]
+            paths.push_back(silo_index_path[path_string][block_id].as_string());
+        }
+    }
+    // leverage RVO
+    return paths;
+}
+
+//-----------------------------------------------------------------------------
+// creates silo tree path generators
+SiloTreePathGenerator*
+create_silo_tree_path_generator(DBfile *root_file,
+                                const int num_domains,
+                                const Node &n_item,
+                                const std::string &what_kind_of_paths)
+{
+    return new SiloTreePathGenerator(
+        // silo database file
+        root_file,
+        
+        // objpath - we are at the root of the root file
+        ".",
+        
+        // number of blocks
+        num_domains,
+        
+        // mesh paths
+        std::move(get_paths(n_item, what_kind_of_paths, num_domains)),
```

Comment:
It would be better to have the class init this vs using a move in the constructor